File: CMPS 109 - C++:Python:JS:Bash:Awk/Lab1/rpnc.py
# ---------------------------------------------------------------------------------
#  LAB 1: Reverse Polish Notation (RPN) Calculator
#  CMPS 109  Spring 2018 University of California Santa Cruz
# 
#  Boothroyd, Nichole
#  nboothro
# 
#  Language:    Python3
#  
#  Description: This program accepts a single string argument in RPN, evaluates it,
#               and prints the answer to stdout.
# 
#  Usage:       python rpnc.py "RPN expression"
# 
# ---------------------------------------------------------------------------------

import sys
import re

stack = []
argS = str(sys.argv[1])             #Save argument string
arg = argS.split()                  #Create a list of the words from string

for each in arg:                    #For each word in the list:
    input = str(each)               #Cast as str for isdigit()
    if input.isdigit():             #Is it a number. . .
        stack.append(each)          #. . . Then push to stack
    else:
        op_2 = int(stack.pop())     #Since it's not a number. . .
        op_1 = int(stack.pop())     #. . . Pop from stack
        if input == "+":            #Evaluate based on operator:
            result = (op_1 + op_2)
        if input is "-":
            result = (op_1 - op_2)
        if input is "/":
            result = (op_1 / op_2)
        if input is "^":
            result = (op_1 ** op_2)
        if input is "*":
           result = (op_1 * op_2)
        stack.append(result)        #Push result onto stack

print(stack.pop())                  #Pop final result from stack

File: CMPS 109 - C++:Python:JS:Bash:Awk/Lab1/rpnc.c
//---------------------------------------------------------------------------------
// LAB 1: Reverse Polish Notation (RPN) Calculator
// CMPS 109  Spring 2018 University of California Santa Cruz
// 
// Boothroyd, Nichole
// nboothro
//
// Language: C
// 
// Description: This program accepts a single string argument in RPN, evaluates it,
//              and prints the answer to stdout. Returns -1 on error, 0 otherwise.
//
// Usage:       gcc rpnc.c -o rpnc
//	        ./rnpc "RPN expression"
//
//---------------------------------------------------------------------------------
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <tgmath.h>

int main(int argc, char *argv[]) {
    int stack[17];
    int sI = 0;                             //stack index
    char * myString = strtok(argv[1], " "); //Pull token off of arg string
    while (myString != NULL)                //Until reaching end of string:
    {
        if (isdigit(*myString)) {           //Check to see if token is a number
            int data = atoi(myString);      //str to int conversion
            stack[sI] = data;               //push onto stack
            sI++;                           //Increment stack index
        }
        else {
            int op_2 = (stack[sI - 1]);     //Since token is not a number. . .
            int op_1 = (stack[sI - 2]);     //Pop values off of stack
            sI -= 2;                        //Update stack index
            int result = -17;               //Dummy value
            
            if (*myString == '+'){          //Evalue based on operator:
                result = (op_1 + op_2);
            }
            else if (*myString == '-'){
                result = (op_1 - op_2);
            }
            else if (*myString == '/'){
                result = (op_1 / op_2);
            }
            else if (*myString == '*'){
                result = (op_1 * op_2);
            }
            else if (*myString == '^'){
                result = (pow(op_1,op_2));
            }
            
            stack[sI] = result;             //push result onto the stack
            sI++;
        }
        myString = strtok (NULL, " ");      //Pull next token off of str
    }

    sI--;                                //"pop" result off the stack
    printf("%d\n", stack[sI]);

    return 0;
}

